src/converter/renderers/render-aem.js

```diff
@@ -14,20 +15,124 @@ import {
   decodeCQMetadata,
   generateHash,
 } from './utils/aem-page-meta-utils.js';
+import { getMatchLanguageForTag } from '../../common/utils/language-utils.js';
 import { getMetadata, setMetadata } from '../modules/utils/dom-utils.js';
-import { writeStringToFileAndGetPresignedURL } from '../../common/utils/file-utils.js';
+import {
+  writeStringToFileAndGetPresignedURL,
+  readFile,
+} from '../../common/utils/file-utils.js';
+import stateLib from '../../common/utils/state-lib-util.js';
 
 export const aioLogger = Logger('render-aem');
 
 const byteSize = (str) => new Blob([str]).size;
 const isLessThanOneMB = (str) => byteSize(str) < 1024 * 1024 - 1024; // -1024 for good measure :)
 
+const mapTagsToTitles = (meta, taxonomyData) => {
+  let locTitles = [];
+  if (!meta?.length) return [];
+  if (!Array.isArray(taxonomyData) || !taxonomyData?.length) {
+    aioLogger.error('Invalid or empty taxonomy data:', taxonomyData);
+    return [];
+  }
+
+  const tags = meta.split(',').map((tag) => tag.trim());
+  locTitles = tags
+    .map((tag) => {
+      const match = taxonomyData.find((item) => item.tag.trim() === tag);
+      return match ? match.title : null;
+    })
+    .filter(Boolean);
+
+  if (locTitles.length) {
+    locTitles = locTitles.join(', ');
+  }
+  return locTitles;
+};
+
+// Fetch tags data from AEM taxonomy sheets
+async function fetchTaxonomyData(pagePath, params, taxonomy) {
+  const state = await stateLib.init();
+  const taxonomyState = await state.get(taxonomy);
+  if (taxonomyState?.value) {
+    aioLogger.debug(`Using cached ${taxonomy}`);
+    try {
+      const cachedData = JSON.parse(taxonomyState.value);
+      if (Array.isArray(cachedData)) return cachedData;
+      aioLogger.warn(`Cached ${taxonomy} is not an array, ignoring.`);
+    } catch (error) {
+      aioLogger.error(`Error parsing cached ${taxonomy} data:`, error);
+    }
+  }
+
+  aioLogger.debug(`Fetching ${taxonomy} data from AEM taxonomy sheet`);
+  const taxonomyPath = `/${taxonomy}.json`;
+  const lang = pagePath.split('/')[1];
+  const language = lang ? getMatchLanguageForTag(lang) : 'default';
+  let result;
+  try {
+    // eslint-disable-next-line no-use-before-define
+    const data = await renderAem(taxonomyPath, params);
+
+    if (data?.headers?.location) {
+      // Handle AEM response larger than 1MB
+      const filePath = data.headers.location;
+      const fileName = path.basename(filePath);
+      const largeResponse = await readFile(fileName);
+      const parsedLargeResponse = JSON.parse(largeResponse);
+      result = parsedLargeResponse?.[language]?.data || [];
+    } else if (data?.body) {
+      const parsedData = JSON.parse(data.body);
+      result = parsedData?.[language]?.data || [];
+    } else {
+      aioLogger.error(`Unexpected response format for ${taxonomy}:`, data);
+      result = [];
+    }
+  } catch (error) {
+    aioLogger.error(`Error fetching ${taxonomy} data:`, error);
+    result = [];
+  }
```

Comment:
You don't want to do this that way.

You need a client for requesting from AEM, a dedicated utility that does the same thing as `renderAEM`

Specifically some of this part here:

```
  const aemURL = `${aemAuthorUrl}/bin/franklin.delivery/${aemOwner}/${aemRepo}/${aemBranch}${path}?wcmmode=disabled`;
  const url = new URL(aemURL);

  const fetchHeaders = { 'cache-control': 'no-cache' };
  if (authorization) {
    fetchHeaders.authorization = authorization;
  }
  if (sourceLocation) {
    fetchHeaders['x-content-source-location'] = sourceLocation;
  }
  ```
  
  Then you want that client to Store the result with io Files.
  
  I'd call this the FranklinServletClient, and give it methods for handling this case as well as the `renderAEM` case
  
  
  Probably something like:
  
  ```
  class FranklinServletClient {
     constructor({
       aemAuthorUrl,
       aemOwner,
       aemRepo,
       aemBranch,
       authorization
    }){
      ....
    }
    
    fetchFromServlet(path, sourceLocation) {
      ... same code in renderAem
    }
    
    fetchAndCache(path, sourceLocation, expiration) {
       ... do fetchFromServlet
       ... store result in IO Files
       ... handle expiration - maybe more advanced ask and can be left out 
    }
  }
  
  
  ```
  
  Then. use `FranklinServletClient#fetchAndCache` for your purpose.

Reply:
Done